numba/listobject.py

```diff
@@ -0,0 +1,1062 @@
+                         likely=True):
+        context.call_conv.return_user_exc(builder, RuntimeError, (msg,))
+
+
+@intrinsic
+def _as_meminfo(typingctx, lstobj):
+    """Returns the MemInfoPointer of a list.
+    """
+    if not isinstance(lstobj, types.ListType):
+        raise TypingError('expected *lstobj* to be a ListType')
+
+    def codegen(context, builder, sig, args):
+        [tl] = sig.args
+        [l] = args
+        # Incref
+        context.nrt.incref(builder, tl, l)
+        ctor = cgutils.create_struct_proxy(tl)
+        lstruct = ctor(context, builder, value=l)
+        # Returns the plain MemInfo
+        return lstruct.meminfo
+
+    sig = _meminfo_listptr(lstobj)
+    return sig, codegen
+
+
+@intrinsic
+def _from_meminfo(typingctx, mi, listtyperef):
+    """Recreate a list from a MemInfoPointer
+    """
+    if mi != _meminfo_listptr:
+        raise TypingError('expected a MemInfoPointer for list.')
+    listtype = listtyperef.instance_type
+    if not isinstance(listtype, ListType):
+        raise TypingError('expected a {}'.format(ListType))
+
+    def codegen(context, builder, sig, args):
+        [tmi, tdref] = sig.args
+        td = tdref.instance_type
+        [mi, _] = args
+
+        ctor = cgutils.create_struct_proxy(td)
+        dstruct = ctor(context, builder)
+
+        data_pointer = context.nrt.meminfo_data(builder, mi)
+        data_pointer = builder.bitcast(data_pointer, ll_list_type.as_pointer())
+
+        dstruct.data = builder.load(data_pointer)
+        dstruct.meminfo = mi
+
+        return impl_ret_borrowed(
+            context,
+            builder,
+            listtype,
+            dstruct._getvalue(),
+        )
+
+    sig = listtype(mi, listtyperef)
+    return sig, codegen
+
+
+@intrinsic
+def _list_set_method_table(typingctx, lp, itemty):
+    """Wrap numba_list_set_method_table
+    """
+    resty = types.void
+    sig = resty(lp, itemty)
+
+    def codegen(context, builder, sig, args):
+        vtablety = ir.LiteralStructType([
+            ll_voidptr_type,  # equal
+            ll_voidptr_type,  # item incref
+            ll_voidptr_type,  # item decref
+        ])
+        setmethod_fnty = ir.FunctionType(
+            ir.VoidType(),
+            [ll_list_type, vtablety.as_pointer()]
+        )
+        setmethod_fn = ir.Function(
+            builder.module,
+            setmethod_fnty,
+            name='numba_list_set_method_table',
+        )
+        dp = args[0]
+        vtable = cgutils.alloca_once(builder, vtablety, zfill=True)
+
+        # install key incref/decref
+        item_equal_ptr = cgutils.gep_inbounds(builder, vtable, 0, 0)
+        item_incref_ptr = cgutils.gep_inbounds(builder, vtable, 0, 1)
+        item_decref_ptr = cgutils.gep_inbounds(builder, vtable, 0, 2)
+
+        dm_item = context.data_model_manager[itemty.instance_type]
+        if dm_item.contains_nrt_meminfo():
+            equal = _get_equal(context, builder.module, dm_item, 'list')
```

Comment:
This will over-restrict the itemty to require the ability to do `==` on it.  This is currently preventing the use of jitclasses, which doesn't support `__eq__`.

Reply:
The equals method in the method table isn't being used anywhere in the C level code and is more of an artifact left over from the copy and edit from the Dict. In could be removed in principle.

Reply:
see https://github.com/esc/numba/pull/5